File: clojure/datamunging/datamunging0.clj
(ns datamunging.datamunging0
  (:use [clojure.string :only [split]]
        clojure.test))

(defn read-file-lines [file-name] (vec (.split (slurp file-name) "\n")))
(defn as-int [s] (. Integer parseInt (.replace s "*" "")))
(defn split-each-line [lines] (map #(vec (.split (.trim %) "\\s+")) lines))
(defn find-row-with-min-diff [data-rows]
  (apply min-key #(Math/abs (- (nth % 1) (nth % 2))) data-rows))


(defn find-day-with-min-temperature-spread []
  (defn filter-non-data-lines [lines] (drop 8 (drop-last 2 lines)))
  (defn parse-as-data [lines] (map #(vec [(nth % 0) (as-int (nth % 1)) (as-int (nth % 2))]) (split-each-line lines)))

  (let [lines (read-file-lines "/Users/dima/IdeaProjects/katas/src/main/scala/ru/katas/n4/weather.dat")]
    (first (find-row-with-min-diff (parse-as-data (filter-non-data-lines lines))))
  )
)

(defn find-team-with-min-goal-diff []
  (defn filter-non-data-lines [lines] (drop-last 1 (drop 5 (filter #(not (.contains % "---")) lines))))
  (defn parse-as-data [lines]
    (map #(vec [(nth % 1) (as-int (nth % 6)) (as-int (nth % 8))])(split-each-line lines)))

  (let [lines (read-file-lines "/Users/dima/IdeaProjects/katas/src/main/scala/ru/katas/n4/football.dat")]
    (first (find-row-with-min-diff (parse-as-data (filter-non-data-lines lines))))
  )
)

(deftest should-find-team-with-min-goal-diff
  (is (= "Aston_Villa" (find-team-with-min-goal-diff)))
)
(deftest should-find-day-with-min-temperature-spread
  (is (= "14" (find-day-with-min-temperature-spread)))
)
(run-tests)

File: scala/src/katas/scala/classic_katas/datamunging/DataMunging0.scala
package katas.scala.classic_katas.datamunging

import org.specs2.matcher.ShouldMatchers
import org.junit.Test
import scala.io.Source

/**
 * User: dima
 * Date: 15/03/2012
 */
class DataMunging0 extends ShouldMatchers {
  @Test def findDayWithMinTemperatureSpread() {
    val lines = trimFile("/Users/dima/IdeaProjects/katas/src/main/scala/ru/katas/n4/weather.dat", 8, 2)
    lines.foreach{ line => println(line) }

    val rows = lines.map{ line =>
        val row = line.split("\\s+")
        val lineNumber = row(1).toInt
        val maxTemperature = row(2).replaceAll("\\*", "").toInt
        val minTemperature = row(3).replaceAll("\\*", "").toInt

        (lineNumber,  MaxMin(maxTemperature, minTemperature))
    }
//    rows.foreach{row( Row => println(row)}

    val rowWithMinSpread = rows.minBy { row => math.abs(row._2.max - row._2.min) }
    println(rowWithMinSpread)

  }

  case class MaxMin(max: Int, min: Int)

//  def smallestDiff(name: String,  m: MaxMin)

  @Test def findATeamWithSmallestDifference() {
    var lines = trimFile("/Users/dima/IdeaProjects/katas/src/main/scala/ru/katas/n4/football.dat", 5, 1)
    lines = lines.filterNot{ _.matches(" *[-]+") }

    val scores = lines.map{ line =>
      val row = line.split("\\s+")
      (row(2), new MaxMin(row(7).toInt, row(9).toInt))
    }
    println(scores.minBy { x => (x._2.max - x._2.min).abs}   )
  }

  def trimFile(file: String, top: Int, bottom: Int): List[String] = {
    val source = Source.fromFile(file)
    var lines = source.getLines().toList
    lines.take(lines.length - bottom).drop(top)
  }
}